File: Data Structures & Algorithms/two-integer-sum/submission-2.py
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        # worst way - iterate with nested for loops to check for combos
        # space optimal
            # sort and use two pointers - (two sum 2 solution)
                # if l + r values > target ; move right ptr inwards (to lower)
                # if < target ; move left ptr to inc
                # if == target ; 
        # most time optimal
            # use a hashmap to quickly [o(1)] look up complements
            # iterate through list
                # calculate complement for each value (target - curr)
                # if it exists in the map (instant lookup), return both indices
                # otherwise, add it to the map (value -> index) 

            # use enumerate to pull index alongside num value 
            # ex. for idx, num in enumerate(nums) 


        hashmap = {}

        for i, num in enumerate(nums):
            complement = target - num
            if complement in hashmap: # this is key because doing "in hashmap" searches the KEYS. 
                # this is why we store the actual values as the map keys and the index in list as the map value
                return [hashmap[complement], i]
                    # we want to return the smaller index first 
                    # the cached values would always have earlier indices
                
            else:
                hashmap[num] = i

        # we are guaranteed 1 valid answer exists

        # TC: O(n) - we iterate through once and map operations are O(1) 
        # SC: O(n) - values stored in the map
        

            

File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-2.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        int l = 0;
        int r = nums.size() - 1;

        while (l <= r) {
            // we want to keep the mid point as a valid

            int m = (l+r) / 2;

            if (nums[m] == target) {
                // we want to return the index
                return m;
            }

            if (nums[l] <= nums[m]) {
                // this is the sorted side
                // check if the target is in the sorted range (aka if it's < the mid)
                if (target > nums[m] || target < nums[l]) {
                    // if so we can search the left half.
                    // move the right pointer
                    l = m + 1; // we move it past m
                    // we do r = m to retain m in search space if not looking for specific value
                }
                else {
                    //move the left pointer
                    r = m - 1;
                }

            }
            else {
                // the right half if sorted

                // check if the target > mid
                if (target < nums[m] || target > nums[r]) {
                    // then we can search the left half
                    r = m - 1;
                }
                else {
                    l = m + 1;
                }
            }
        }

        return -1; // edgecase
    }
};
